src/simplify_reshapes.cpp

```diff
@@ -318,11 +318,224 @@ struct find_nested_concat
+    }
+
+    void apply(module& p, match::matcher_result r) const
+    {
+        auto ins     = r.result;
+        auto ins_rsp = r.instructions["data"];
+        auto ins_ind = r.instructions["ind"];
+
+        // resize input shape
+        if(ins_rsp->get_shape().lens().size() != 1)
+        {
+            return;
+        }
+
+        // resize output shape
+        const auto& in_shape  = ins_rsp->inputs().front()->get_shape();
+        const auto& out_shape = ins->get_shape();
+        // check if output shape is multiple of input shape
+        const auto& in_lens  = in_shape.lens();
+        const auto& out_lens = out_shape.lens();
+        if(in_lens.size() != out_lens.size())
+        {
+            return;
+        }
+
+        // output shape must be multiple of input shape
+        std::vector<bool> is_multi(in_lens.size());
+        std::transform(
+            in_lens.begin(), in_lens.end(), out_lens.begin(), is_multi.begin(), [](auto x, auto y) {
+                return (y % x == 0);
+            });
+        if(not std::all_of(is_multi.begin(), is_multi.end(), [](auto b) { return b; }))
+        {
+            return;
+        }
+
+        // output must be multiple of inputs
+        std::vector<std::size_t> scales(in_lens.size());
+        std::transform(
+            in_lens.begin(), in_lens.end(), out_lens.begin(), scales.begin(), [](auto x, auto y) {
+                return y / x;
+            });
+
+        // if ind is not constant, cannot optimize
+        std::vector<int> vec_ind;
+        auto arg_ind = ins_ind->eval();
+        if(arg_ind.empty())
+        {
+            return;
+        }
+        arg_ind.visit([&](auto v) { vec_ind.assign(v.begin(), v.end()); });
+        std::vector<int> index(out_shape.elements());
+        std::iota(index.begin(), index.end(), 0);
+        if(not std::all_of(index.begin(), index.end(), [&](auto i) {
+               auto out_idx = out_shape.multi(i);
+               auto in_idx  = out_idx;
+               std::transform(out_idx.begin(),
+                              out_idx.end(),
+                              scales.begin(),
+                              in_idx.begin(),
+                              [&](auto io, auto scale) { return io - (io % scale); });
+               return vec_ind[i] == vec_ind[out_shape.index(in_idx)];
+           }))
+        {
+            return;
+        }
+
+        // wrap up shapes for multibroadcast
+        std::vector<int64_t> in_dims(in_lens.begin(), in_lens.end());
+        std::vector<int64_t> out_dims(out_lens.begin(), out_lens.end());
+        for(int ii = static_cast<int>(scales.size()) - 1; ii >= 0; --ii)
+        {
+            if(scales[ii] == 1 or in_dims[ii] == 1)
+            {
+                continue;
+            }
+
+            out_dims[ii] = in_dims[ii];
+            in_dims.insert(in_dims.begin() + ii + 1, 1);
+            out_dims.insert(out_dims.begin() + ii + 1, scales[ii]);
+        }
```

Comment:
It seems like it would be better to zip the scales and dims together:

```cpp
std::vector<std::pair<std::size_t, std::size_t>> dim_scales;
std::transform(in_lens.begin(), in_lens.end(), out_lens.begin(), dim_scales.begin(), [](auto x, auto y) {
    return std::make_pair(x, y / x);
});
```

Then you could create the vector of dims by a simple iteration:

```cpp
for(auto&& p:dim_scales)
{
    in_dims.push_back(p.first);
    out_dims.push_back(p.first);
    if(p.first == 1 or p.second == 1)
        continue;
    in_dims.push_back(1);
    out_dims.push_back(p.second);
}
```

This would also avoid inserting in the middle of the vector since that is slow.

Reply:
fixed